File: GradualConstraints/src/de/unifreiburg/cs/proglang/jgs/jimpleutils/Supertypes.java
package de.unifreiburg.cs.proglang.jgs.jimpleutils;

import soot.RefType;
import soot.SootClass;
import soot.SootMethod;
import soot.Type;

import java.sql.Ref;
import java.util.*;
import java.util.function.Function;
import java.util.stream.Collector;
import java.util.stream.Collectors;
import java.util.stream.Stream;
import java.util.stream.StreamSupport;

import static java.util.Arrays.asList;
import static java.util.stream.Collectors.toList;
import static java.util.stream.Collectors.toSet;

public class Supertypes {

    /**
     * @return The supertypes of {@code sootClass}. Traverses the supertype DAG in preorder. The stream may contain
     * duplicates.
     */
    //TODO-performance: could avoid traversing classes multiple times (but how large can the hierarchies be?)
    public static Stream<SootClass> enumerate(SootClass sootClass) {
        List<SootClass> immediateSuperTypes =
                Stream.concat(
                  sootClass.hasSuperclass()
                  ? Stream.of(sootClass.getSuperclass())
                  : Stream.empty(),
                sootClass.getInterfaces().stream()).collect(toList());
        return Stream.concat(immediateSuperTypes.stream(), immediateSuperTypes.stream().flatMap(Supertypes::enumerate));
    }

    /**
     * @return The methods that m1 overrides.
     */
    public static Stream<SootMethod> findOverridden(SootMethod m1) {
        SootClass c1 = m1.getDeclaringClass();
        return Supertypes.enumerate(c1)
                .flatMap(c -> c.getMethods().stream())
                .filter(m -> overrides(m1, m));
    }

    /**
     * @return true if {@code m1} overrides {@code m2}.
     */
    public static boolean overrides(SootMethod m1, SootMethod m2) {
        // ignore static methods and constructors
        if (m1.isStatic() || m2.isStatic()) { return false; }
        if (m1.getName().equals("<init>") || m2.getName().equals("<init>")) { return false; }
        if (m1.getName().equals(m2.getName())
                && subTypeOf(m1.getDeclaringClass(), m2.getDeclaringClass())
                && m1.getParameterTypes().equals(m2.getParameterTypes())) {
            // it remains to check the return type
            Type rt1 = m1.getReturnType();
            Type rt2 = m2.getReturnType();
            if (rt1 instanceof RefType && rt2 instanceof RefType) {
                return subTypeOf(((RefType)rt1).getSootClass(), ((RefType)rt2).getSootClass());
            } else {
               return rt1.equals(rt2);
            }
        } else {
            return false;
        }
    }

    /**
     * @return true if {@code c1} is a subtype of {@code c2}
     */
    public static boolean subTypeOf(SootClass c1, SootClass c2) {
        return Supertypes.enumerate(c1).filter(sc1 -> sc1.equals(c2)).findAny().isPresent();
    }
}

File: GradualConstraints/src/de/unifreiburg/cs/proglang/jgs/jimpleutils/Supertypes.scala
package de.unifreiburg.cs.proglang.jgs.jimpleutils

import soot.RefType
import soot.SootClass
import soot.SootMethod
import soot.Type

import scala.collection.JavaConversions._

object Supertypes {
  /**
    * @return The supertypes of { @code sootClass}. Traverses the supertype DAG in preorder. The stream may contain
    *                                   duplicates.
    */
  def enumerate(sootClass: SootClass): Iterator[SootClass] = {
    val immediateSuperTypes: List[SootClass] = (if (sootClass.hasSuperclass) List(sootClass.getSuperclass) else List()) ++ sootClass.getInterfaces
    immediateSuperTypes.iterator ++ immediateSuperTypes.flatMap(Supertypes.enumerate)
  }

  /**
    * @return The methods that m1 overrides.
    */
  def findOverridden(m1: SootMethod): Iterator[SootMethod] = {
    val c1: SootClass = m1.getDeclaringClass

    //Supertypes.enumerate(c1).flatMap(c -> c.getMethods().stream()).filter(m -> overrides(m1, m))
    for (c <- Supertypes.enumerate(c1);
         m <- c.getMethods if overrides(m1, m))
      yield m
  }

  /**
    * @return true if { @code m1} overrides { @code m2}.
    */
  def overrides(m1: SootMethod, m2: SootMethod): Boolean = {
    if (m1.isStatic || m2.isStatic) {
      return false
    }
    if ((m1.getName == "<init>") || (m2.getName == "<init>")) {
      return false
    }
    if ((m1.getName == m2.getName) && subTypeOf(m1.getDeclaringClass, m2.getDeclaringClass) && (m1.getParameterTypes == m2.getParameterTypes)) {
      val rt1: Type = m1.getReturnType
      val rt2: Type = m2.getReturnType
      if (rt1.isInstanceOf[RefType] && rt2.isInstanceOf[RefType]) {
        return subTypeOf((rt1.asInstanceOf[RefType]).getSootClass, (rt2.asInstanceOf[RefType]).getSootClass)
      }
      else {
        return rt1 == rt2
      }
    }
    else {
      return false
    }
  }

  /**
    * @return true if { @code c1} is a subtype of { @code c2}
    */
  def subTypeOf(c1: SootClass, c2: SootClass): Boolean = Supertypes.enumerate(c1).contains(c2)
}